ostd/src/task/scheduler/mod.rs

```diff
@@ -0,0 +1,273 @@
+        if let Some(next) = local_rq.pick_next_current() {
+            ReschedAction::SwitchTo(next)
+        } else {
+            ReschedAction::Retry
+        }
+    })
+}
+
+/// Blocks the current task unless `has_woken` is `true`.
+pub(crate) fn park_current(has_woken: &AtomicBool) {
+    let mut current = None;
+    let mut is_first_try = true;
+    reschedule(&mut |local_rq: &mut dyn LocalRunQueue| {
+        if is_first_try {
+            if has_woken.load(Ordering::Acquire) {
+                return ReschedAction::DoNothing;
+            }
+            current = local_rq.dequeue_current();
+            local_rq.update_current(UpdateFlags::Wait);
+        }
+        if let Some(next_task) = local_rq.pick_next_current() {
+            if Arc::ptr_eq(current.as_ref().unwrap(), &next_task) {
+                return ReschedAction::DoNothing;
+            }
+            ReschedAction::SwitchTo(next_task)
+        } else {
+            is_first_try = false;
+            ReschedAction::Retry
+        }
+    });
+}
+
+/// Unblocks a target task.
+pub(crate) fn unpark_target(runnable: Arc<Task>) {
+    let should_preempt_info = SCHEDULER
+        .get()
+        .unwrap()
+        .enqueue(runnable, EnqueueFlags::Wake);
+    if should_preempt_info.is_some() {
+        let cpu_id = should_preempt_info.unwrap();
+        // FIXME: send IPI to set remote CPU's `SHOULD_PREEMPT` if needed.
+        if cpu_id == this_cpu() {
+            SHOULD_PREEMPT.set();
+        }
+    }
+}
+
+/// Preempts the current task.
+pub(crate) fn preempt() {
+    fn preempt_check() -> bool {
+        PREEMPT_INFO.load() == 0
+    }
+
+    if !preempt_check() {
+        return;
+    }
+    yield_now();
+}
+
+/// Yields execution.
+pub(super) fn yield_now() {
+    reschedule(&mut |local_rq| {
+        local_rq.update_current(UpdateFlags::Yield);
+
+        if let Some(next) = local_rq.pick_next_current() {
+            ReschedAction::SwitchTo(next)
+        } else {
+            ReschedAction::DoNothing
+        }
+    })
+}
+
+/// Do rescheduling by acting on the scheduling decision (`ReschedAction`) made by a
+/// user-given closure.
+///
+/// The closure makes the scheduling decision by taking the local runqueue has its input.
+fn reschedule<F>(f: &mut F)
+where
+    F: FnMut(&mut dyn LocalRunQueue) -> ReschedAction,
+{
+    let next_task = loop {
+        let mut action = ReschedAction::DoNothing;
+        SCHEDULER.get().unwrap().local_mut_rq_with(&mut |rq| {
+            action = f(rq);
+        });
+
+        match action {
+            ReschedAction::DoNothing => {
+                return;
+            }
+            ReschedAction::Retry => {
+                continue;
+            }
+            ReschedAction::SwitchTo(next_task) => {
+                break next_task;
+            }
+        };
+    };
+
+    SHOULD_PREEMPT.clear();
+    processor::switch_to_task(next_task);
+}
+
+/// A per-CPU task scheduler.
+pub trait Scheduler<T = Task>: Sync + Send {
```

Comment:
A readable Rust module should be read like a news paper: high-level and important items are put at the beginning of a module and low-level and unimportant items at the end.

Thus, move all public types such as `Scheduler` to the top of the file, close to `inject_scheduler`. Public APIs are important than crate-level public APIs, the latter of which are more important than private items.